ra/ra.go

```diff
@@ -1844,23 +1848,56 @@ func (ra *RegistrationAuthorityImpl) RevokeCertByApplicant(ctx context.Context,
 		logEvent.Reason = req.Code
 	}
 
-	issuerID := issuance.IssuerNameID(cert)
 	err = ra.revokeCertificate(
 		ctx,
-		cert.SerialNumber,
-		issuerID,
+		cert,
 		revocation.Reason(req.Code),
 	)
 	if err != nil {
 		return nil, err
 	}
 
 	// Don't propagate purger errors to the client.
+	issuerID := issuance.IssuerNameID(cert)
 	_ = ra.purgeOCSPCache(ctx, cert, issuerID)
 
 	return &emptypb.Empty{}, nil
 }
 
+// crlShard extracts the CRL shard from a certificate's CRLDistributionPoint.
+//
+// If there is no CRLDistributionPoint, returns 0.
+//
+// If there is more than one CRLDistributionPoint, returns an error.
+//
+// Assumes the shard number is represented in the URL as an integer that
+// occurs in the last path component, optionally followed by ".crl".
```

Comment:
I admit that this is the place that had me agonizing the most of what the "right thing to do" is when I was thinking about this six months ago. Your approach here is nice and flexible, in that it would allow us to change our CRLDP "base" (e.g. the hostname or first part of the path) without having to change this code. But it's also somewhat scary in that it will happily process a certificate with a very bizarre CRLDP and not complain.

I think the approach you've settled on here is probably the best path forward for now -- the main alternative I see is configuring the crldpBaseUrl in the RA as well as in the CA, and that's a deployability mess that has only become more complex with the advent of multiple issuers -- but I think we should explicitly document this potential footgun.

Reply:
Yep, I agree on the scariness; but I think we're better off catching bizarre CRLDPs at linting time (which covers all certs, and can catch problems in advance) instead of revocation time.